flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -322,23 +350,71 @@ private MemorySegment requestMemorySegment() {
 		return requestMemorySegment(UNKNOWN_CHANNEL);
 	}
 
-	@Nullable
-	private MemorySegment requestMemorySegmentFromGlobal() {
-		assert Thread.holdsLock(availableMemorySegments);
+	private boolean requestMemorySegmentFromGlobal() {
+		if (numberOfRequestedMemorySegments >= currentPoolSize) {
+			return false;
+		}
+
+		MemorySegment segment = networkBufferPool.requestMemorySegment();
+		if (segment != null) {
+			availableMemorySegments.add(segment);
+			numberOfRequestedMemorySegments++;
+			return true;
+		}
+		return false;
+	}
```

Comment:
maybe change it to `while (numberOfRequestedMemorySegments < currentPoolSize)` loop? (as a follow up commit?)

Reply:
Is your intent to reach equilibrium much quicker? If so, I like the idea. (If not, then I haven't understood)

My main concern is that it means that the first buffer pool would potentially take all available segments while the last buffer pool gets nothing although each of them could take some buffers. However, I must admit that I have not fully understood when excess buffers actually occur in reality. I'd assume that during start of an application all pools / exclusive segments are more or less acquired simultaneously and handed out a bit later to the writer/input channels, such that excess buffers are close to non-existant.

Reply:
As discussed offline, let's keep it as it is for now.